micheline/script.go

```diff
@@ -157,11 +157,47 @@ func (s Script) Bigmaps() map[string]int64 {
 	return DetectBigmaps(s.Code.Storage, s.Storage)
 }
 
+// Flattens pair primitives. Basically the same as the UNPAIR michelson operation.
+// For instance, `pair (pair (pair 1 2) 3) 4` becomes `pair 1 2 3 4`. Other container
+// primitives like map and list remain untouched.
+func flatten(p Prim) []Prim {
```

Comment:
This is needed because storage values are not always in the same shape as the type definitions in the storage code, specifically for pairs. Nested pair definitions can actually accept flattened storage values (see contract `KT1CSYNJ6dFcnsV4QJ6HnBFtdif8LJGPQiDM`,) so I flatten all pairs here just for consistency. In function `linkStorageTypeAndValue`, because pair values are assumed to be flattened, pair primitives continue to use the same value queue, unlike maps and lists that unfold the value on the spot and proceed recursively with the new unfolded value queue.